### DISCUSSION TITLE
pybricks-mpy-cross: does this work for bricks other than ev3?

### DISCUSSION THREAD

**Discussion by jaguilar** [pybricks/support#1451] [Q&A] [answered]
**Sorry, the title is not quite correct -- see TL;DR for my actual question.**

Tl;dr: Do [mpy files containing native code](https://docs.micropython.org/en/latest/develop/natmod.html) work in pybricks on bricks other than ev3?

Hey there, this is just a quick curiosity type of question. I'm working on speeding up some of my code, and I'm not sure if it makes sense to keep hacking around with Viper, or if I should rewrite the hot part of the loop in C++. I learned that mpy files can be used to package native code for use with micropython, and I saw that pybricks-mpy-cross exists, so I assumed that ev3 could use mpy files with hand-rolled native code. So, that's cool! If my assumption is correct, I can make a C++ version of my PID controller and call into that on ev3dev.

However! I would prefer not to use techniques that won't work on more recent lego bricks. That's why I'm using Pybricks, after all, instead of ev3dev C++. So my question then is simple: is including native code in an mpy file a technique that works across all lego bricks, or is it only available for ev3?

**Comment by jaguilar** [comment]
Partly answering my own question: the ability to add native code as an mpy file is apparently more recent than the version of micropython used on ev3dev. At least, if the lack of `py/dynruntime.h` [in the source code](https://github.com/pybricks/micropython/tree/ev3dev-stretch/py) is any indication. So, on EV3, I think that the `ffi` module might be my only play for importing native code into pybricks-micropython. Please let me know if I'm missing something!

**Comment by dlech** [comment, marked as answer]
It is supported on SPIKE and Robot Inventor hubs since last year (https://github.com/pybricks/pybricks-micropython/pull/149). You will need an mpy-cross version that matches the MicroPython version the Pybricks firmware is based on. The EV3 version is too old for the newer hubs. One way to get this is `pipx install mpy-cross-v6.1`. Or you can build your own from the Pybricks source code.

For EV3, yes, `ffi` would be the way to go.

**Reply by jaguilar** [reply]
I'm interested in adding an example of this to the pybricks-api project. It's proven quite challenging for me to get working so I imagine an example would be helpful for others. One of the things about cross-compiling is that a lot of us embedded systems neophytes have no experience with it, and don't even know the basic principles without a bunch of research.

It occurred to me that probably the very easiest way of getting a working binary is to use the same docker image as y'all use to build pybricks for ev3. I managed to get that working on my machine, but I've run out of time and will be out of town for the next week.

My question is: would y'all want to have such an example in the repo? Would referring to your ev3dev build environment docker image be an acceptable way to cross-compile? And if so, what's the best way to refer to that docker image? Just the URL? Or instruct the user to clone the pybricks-micropython repo and provide a path?

**Reply by dlech** [reply]
A github repo with a "hello world" example would probably be the most useful. I would expect this to have a Dockerfile in it that uses the docker image and adds any extras needed and mounts a local volume with the source code from the repo.

**Reply by jaguilar** [reply]
Assuming I can get it working, I will publish that. Thanks! :)
